_posts/2017-03-02-lang-ergonomics.md

```diff
@@ -0,0 +1,408 @@
+on `K`.
+
+* Trade power/precision: this is a feature
+that allows for quite widespread elision. But in practice, what's being elided
+tends to be well-known, and even when it's not, it tends not to cause too much
+trouble. After all, when *using* a function like `use_map`, you're generally
+going to be passing in an existing `HashMap`, which by construction will ensure
+that the bounds already hold.
+
+* Limit context: it's straightforward to discover how the bounds are being
+imposed by examining the type definitions, and the compiler can reliably produce
+an error pointing directly to the type(s) imposing unfulfilled bounds.
+
+## Example: ownership
+
+A lot of work went into making Rust's ownership system ergonomic, and that work
+entailed judicious use of "implicit" features. It's particularly instructive to
+look at the places where borrowing is explicit, and places where it's not:
+
+- Borrowing is implicit for the receiver when invoking a method.
+- Borrowing is explicit for normal function arguments and in other expressions.
+
+This design was arrived at after many iterations with alternatives, and it's
+turned out quite nicely. It's another example of the power/precision
+tradeoff. We provide powerful borrowing inference but only for a narrowly
+limited location.
+
+Ownership is important in Rust, and reasoning locally about it is
+vital. However:
+
+- The vast majority of methods borrow `self`, rather than taking it by value.
+- Usually the name of a method makes obvious whether the `self` borrow will be
+  mutable or not (e.g. `push` versus `len`).
+- The borrow checker will prevent any incorrect borrowing, though if borrowing
+  were too implicit, it could make debugging borrow check errors harder.
+
+Neither of the first two points apply as strongly to method/function
+arguments. So we ended up with a system that is neither fully explicit nor fully
+implicit, but rather one that balances good ergonomics with a compact reasoning
+footprint. **This design also aids learnability, by often just doing "the
+obvious thing" for borrowing, and thereby limiting the situations in which
+newcomers have to grapple with choices about it**.
+
+### Ideas: implied borrows
+
+Nevertheless, there are some pain points around borrowing in Rust today. To wit:
+
+**Discarding ownership**. Sometimes you have ownership of a value, like a
+`String`, and want to pass it to a function that only needs a borrow (say,
+`&str`), after which you no longer need the value. Today, you *must* borrow the
+value in the argument:
+
+```rust
+fn read_config(path: &Path) -> Config { ... }
+
+let mut path = PathBuf::new(src_dir);
+path.push("Config.toml");
+
+// we have to borrow `path` with `&` even though we're done with it
+let config = read_config(&path);
+```
+
+But we could easily allow you to write `read_config(path)`, implicitly borrowing
+`path` for `read_config` and then *dropping it* immediately afterwards. That
```

Comment:
I still feel like it might be helpful to say, effectively desugaring into something like:

```rust
let config = {
    let tmp = path;
    read_config(&tmp)
};
```

but perhaps the syntax is too fancy